### PR TITLE
Add DNS Provider for Tencent Cloud

### PR DESCRIPTION
**Pull request by baranwang**
Add DNS Provider for Tencent Cloud

### CODE REVIEW

**Comment by ldez** [comment]
Hello, in order for a PR adding a DNS provider to be accepted, you have to:

- [ ] add a description to your PR
- [ ] be able to maintain this provider
- [ ] have a homogeneous design with the other providers
- [ ] add tests (units)
```shell
make test
```
- [ ] add tests ("live")
https://github.com/go-acme/lego/blob/c938de68f262c057542d44bcb67ef961cefcc60e/providers/dns/cloudflare/cloudflare_test.go#L125-L151

```shell
make test
```
- [ ] add a [provider descriptor](https://github.com/go-acme/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [ ] generate CLI help, documentation, and readme.
```bash
make generate-dns
```
- [ ] be able to do: _(and put the output of this command to a comment in your PR)_
```bash
rm -rf .lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.
- [ ] pass the linter ([golangci-lint](https://github.com/golangci/golangci-lint#install) must be installed):
```shell
make checks
```
- [ ] do `go mod tidy`

**Comment by baranwang** [comment]
```
go run cmd/lego/main.go -m your@email.com --dns tencentcloud -d test.neobaran.com -s https://acme-staging-v02.api.letsencrypt.org/directory run      
2021/11/09 15:00:12 [INFO] [test.neobaran.com] acme: Obtaining bundled SAN certificate
2021/11/09 15:00:13 [INFO] [test.neobaran.com] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/894627238
2021/11/09 15:00:13 [INFO] [test.neobaran.com] acme: Could not find solver for: tls-alpn-01
2021/11/09 15:00:13 [INFO] [test.neobaran.com] acme: Could not find solver for: http-01
2021/11/09 15:00:13 [INFO] [test.neobaran.com] acme: use dns-01 solver
2021/11/09 15:00:13 [INFO] [test.neobaran.com] acme: Preparing to solve DNS-01
2021/11/09 15:00:13 [INFO] [test.neobaran.com] acme: Trying to solve DNS-01
2021/11/09 15:00:13 [INFO] [test.neobaran.com] acme: Checking DNS record propagation using [198.18.0.2:53]
2021/11/09 15:00:15 [INFO] Wait for propagation [timeout: 10m0s, interval: 2s]
2021/11/09 15:00:15 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:17 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:19 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:21 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:23 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:25 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:27 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:29 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:31 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:33 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:35 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:37 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:39 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:41 [INFO] [test.neobaran.com] acme: Waiting for DNS record propagation.
2021/11/09 15:00:51 [INFO] [test.neobaran.com] The server validated our request
2021/11/09 15:00:51 [INFO] [test.neobaran.com] acme: Cleaning DNS-01 challenge
2021/11/09 15:00:51 [INFO] [test.neobaran.com] acme: Validations succeeded; requesting certificates
2021/11/09 15:00:57 [INFO] [test.neobaran.com] Server responded with a certificate.
```

**Comment by baranwang** [comment]
i try master branch, some error, so i think the unit test can not passed not my request's problem.

**Comment by ldez** [comment]
The failures come from the tests of your provider, so it's a problem with your PR.

File: providers/dns/tencentcloud/tencentcloud_test.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+		expected  string
+	}{
+		{
+			desc:      "success",
+			secretID:  "123",
+			secretKey: "456",
+		},
+		{
+			desc: "success",
+		},
```
Comment (ldez) [inline providers/dns/tencentcloud/tencentcloud_test.go:90]:
```suggestion
```

File: providers/dns/tencentcloud/tencentcloud.toml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+Name = "Tencent Cloud DNS"
+Description = ''''''
+URL = "https://cloud.tencent.com/product/cns"
+Code = "tencentcloud"
+Since = "v4.5.4"
```
Comment (ldez) [inline providers/dns/tencentcloud/tencentcloud.toml:5]:
```suggestion
Since = "v4.6.0"
```

File: providers/dns/tencentcloud/tencentcloud.toml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+TENCENTCLOUD_SECRET_KEY = "Access Key secret"
+[Configuration.Additional]
+TENCENTCLOUD_POLLING_INTERVAL = "Time between DNS propagation check"
+TENCENTCLOUD_PROPAGATION_TIMEOUT = "Maximum waiting time for DNS propagation"
+TENCENTCLOUD_TTL = "The TTL of the TXT record used for the DNS challenge"
+TENCENTCLOUD_HTTP_TIMEOUT = "API request timeout"
+
+[Links]
+API = "https://cloud.tencent.com/document/product/1427/56153"
+GoClient = "https://github.com/tencentcloud/tencentcloud-sdk-go"
```
Comment (ldez) [inline providers/dns/tencentcloud/tencentcloud.toml:25]:
```suggestion
[Configuration]
  [Configuration.Credentials]
    TENCENTCLOUD_SECRET_ID = "Access key ID"
    TENCENTCLOUD_SECRET_KEY = "Access Key secret"
  [Configuration.Additional]
    TENCENTCLOUD_POLLING_INTERVAL = "Time between DNS propagation check"
    TENCENTCLOUD_PROPAGATION_TIMEOUT = "Maximum waiting time for DNS propagation"
    TENCENTCLOUD_TTL = "The TTL of the TXT record used for the DNS challenge"
    TENCENTCLOUD_HTTP_TIMEOUT = "API request timeout"

[Links]
  API = "https://cloud.tencent.com/document/product/1427/56153"
  GoClient = "https://github.com/tencentcloud/tencentcloud-sdk-go"
```

File: providers/dns/tencentcloud/tencentcloud.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+	return NewDNSProviderConfig(config)
+}
+
+// NewDNSProviderConfig return a DNSProvider instance configured for alidns.
+func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
+	if config == nil {
+		return nil, errors.New("tencentcloud: the configuration of the DNS provider is nil")
+	}
+
```
Comment (ldez) [inline providers/dns/tencentcloud/tencentcloud.go:81]:
```suggestion

	if config.SecretID == "" || config.SecretKey == "" {
		return nil, errors.New("tencentcloud: credentials missing")
	}

```

**Review by ldez — CHANGES_REQUESTED** [review]

**Comment by ldez** [comment]
Hello,

could you reopen the PR with a personal fork?
GitHub doesn't allow to change/edit of the content of a PR from an organization.

**Comment by ldez** [comment]
ping @baranwang

**Comment by ldez** [comment]
ping @baranwang

**Comment by ldez** [comment]
last call @baranwang

**Comment by baranwang** [comment]
@ldez sorry, I'm busy at work recently, didn't see it.

**Comment by ldez** [comment]
https://github.com/go-acme/lego/pull/1527#issuecomment-968312246

**Comment by baranwang** [comment]
> [#1527 (comment)](https://github.com/go-acme/lego/pull/1527#issuecomment-968312246)

I have transferred the warehouse to my personal account

**Review by ldez — APPROVED** [review]
Thanks

**Comment by libla** [comment]
I use this providers but throw error:
{"errorCode":1,"errorMessage":"user code exception caught","stackTrace":"invoke function failed with err: {\"Message\":\"error: one or more domains had a problem:\\n[*.honghuangyouxi.com] [*.honghuangyouxi.com] acme: error presenting token: tencentcloud: create record failed: [TencentCloudSDKError] Code=InvalidParameter.DomainInvalid, Message=域名不正确，请输入主域名，如 dnspod.cn。, RequestId=402891c3-cfff-45c0-9248-bfe006aab0bf\\n[honghuangyouxi.com] [honghuangyouxi.com] acme: error presenting token: tencentcloud: create record failed: [TencentCloudSDKError] Code=InvalidParameter.DomainInvalid, Message=域名不正确，请输入主域名，如 dnspod.cn。, RequestId=74f28883-f037-4a90-876f-1e1da90cc6a6\\n\",\"Type\":\"obtainError\",\"StackTrace\":null,\"ShouldExit\":false}","statusCode":430}

Is it because of the wildcard domain?

**Comment by ldez** [comment]
ping @baranwang

**Comment by charlie0129** [comment]
> I use this providers but throw error:
> 
> ```
> {"errorCode":1,"errorMessage":"user code exception caught","stackTrace":"invoke function failed with err: {\"Message\":\"error: one or more domains had a problem:\\n[*.honghuangyouxi.com] [*.honghuangyouxi.com] acme: error presenting token: tencentcloud: create record failed: [TencentCloudSDKError] Code=InvalidParameter.DomainInvalid, Message=域名不正确，请输入主域名，如 dnspod.cn。, RequestId=402891c3-cfff-45c0-9248-bfe006aab0bf\\n[honghuangyouxi.com] [honghuangyouxi.com] acme: error presenting token: tencentcloud: create record failed: [TencentCloudSDKError] Code=InvalidParameter.DomainInvalid, Message=域名不正确，请输入主域名，如 dnspod.cn。, RequestId=74f28883-f037-4a90-876f-1e1da90cc6a6\\n\",\"Type\":\"obtainError\",\"StackTrace\":null,\"ShouldExit\":false}","statusCode":430}
> ```
> 
> Is it because of the wildcard domain?

I ran into this issue too. There is something wrong in the implementation, i.e. the trailing dot in `domainData.domain` is causing problems. I will (probably) provide a fix.

**Pull request merged**